crates/datafusion/src/physical_plan/clickhouse.rs

```diff
@@ -139,172 +140,258 @@ impl BuiltinScalarFunction {
             | BuiltinScalarFunction::ToQuarter => {
                 Signature::Uniform(1, vec![DataType::Date16, DataType::Timestamp32(None)])
             }
-            BuiltinScalarFunction::ToDate => Signature::Uniform(1, vec![DataType::Utf8]),
+            BuiltinScalarFunction::ToDate => Signature::OneOf(vec![
+                Signature::Uniform(
+                    1,
+                    vec![DataType::Date16, DataType::Timestamp32(None)],
+                ),
+                Signature::Uniform(1, vec![DataType::Date16, DataType::LargeUtf8]),
+                Signature::Uniform(1, vec![DataType::Date16, DataType::Int64]),
+            ]),
             BuiltinScalarFunction::ToHour
             | BuiltinScalarFunction::ToMinute
-            | BuiltinScalarFunction::ToSecond => Signature::Uniform(1, vec![DataType::Timestamp32(None)]),
+	    | BuiltinScalarFunction::ToSecond => Signature::Uniform(1, vec![DataType::Timestamp32(None)]),
             BuiltinScalarFunction::EndsWith => Signature::Any(2),
         }
     }
 }
 
 /// Extracts the years from Date16 array
-pub fn date16_to_year(array: &Date16Array) -> Result<UInt16Array> {
-    Ok(array.iter().map(|x| Some(days_to_year(x? as i32))).collect())
+#[inline]
+fn date16_to_year(x: &u16) -> Result<Option<u16>> {
```

Comment:
two bad smells:
1. passing a integer pointer but we just want the value.
2. the original function is column-wise, and so is friendly to testability and performance tuning. you change wrapping batch logics into the macro.

Reply:
Yeah, and you're right. My initial idea was to be clear about the internal native types so that I could perform a single line of performance optimization independently, I will move out of the batch logic from the macro.